language/move-core/types/src/language_storage.rs

```diff
@@ -79,6 +103,153 @@ impl<'a> From<&'a TypeTag> for SafeTypeTag<'a> {
     }
 }
 
+// u8 -> depth or nesting level
+struct TypeTagVisitor(u8);
+
+use serde::de::{
+    self, DeserializeSeed, Deserializer, EnumAccess, MapAccess, SeqAccess, VariantAccess, Visitor,
+};
+
+impl<'de> Visitor<'de> for TypeTagVisitor {
+    type Value = TypeTag;
+
+    fn expecting(&self, formatter: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        formatter.write_fmt(format_args!("`{}`", TYPE_TAG_VARIANTS[0]))?;
+        for value in &TYPE_TAG_VARIANTS[1..] {
+            formatter.write_fmt(format_args!(", `{}`", value))?;
+        }
+        Ok(())
+    }
+
+    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
+    where
+        A: EnumAccess<'de>,
+    {
+        let (variant_type, variant) = data.variant::<DeTypeTag>()?;
```

Comment:
How do we get a match here of `DeTypeTag` vs the actual type? By position in the variant list of `enum`? If so it would be helpful if that has been documented with a short sentence at `DeTypeTag`.